snuba/utils/streams/abstract.py

```diff
@@ -158,18 +176,26 @@ def tell(self) -> Mapping[TStream, TOffset]:
     @abstractmethod
     def seek(self, offsets: Mapping[TStream, TOffset]) -> None:
         """
-        Change the read offsets for the provided streams.
+        Change the working offsets for the provided streams.
+
+        Raises a ``RuntimeError`` if called on a closed consumer.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def stage_offsets(self, offsets: Mapping[TStream, TOffset]) -> None:
+        """
+        Stage offsets for the next commit.
 
         Raises a ``RuntimeError`` if called on a closed consumer.
         """
         raise NotImplementedError
 
     @abstractmethod
-    def commit(self) -> Mapping[TStream, TOffset]:
+    def commit_offsets(self) -> Mapping[TStream, TOffset]:
```

Comment:
So this implies that the normal way to use this api is:
-poll
-poll
-more poll
-stage
-commit
-poll
-....

What was the reasoning to make staging process explicitly required (since the confluent api does not require manually staging the offset) ? And should we call it out explicitly in the doc since the confluent api behaves differently by default ?

Reply:
Yeah, this is a bit confusing and/or misleading.

Right now, the abstract interface doesn't make any promises about whether or not the implementation of the interface itself will or can automatically stage offsets for the messages returned from `poll`. The `KafkaConsumer` implementation _does_ support this by configuration (mostly because that was already in the existing configuration it inherited from `confluent_kafka`/`librdkafka`) but there should probably be some defined generic behavior that is expected from the implementation, and anything that is applicable to any and all backends beyond the bare minimum implementation details (such as automatic offset staging for example) should happen in a class that wraps the abstract one.

CDC is a good example of this approach: https://github.com/getsentry/cdc/blob/1f66b850812365d02b9a9a7942d5f10b4cbe4791/cdc/sources/__init__.py

Reply:
Makes sense to me. Do you plan to make any change ?

Reply:
I think so, but those will be part of #556 — this will need to be rebased on top of those changes when those are ready. Sorry for all of the branch surgery — it seems like one thing always leads to another with these changes…